tests/integration/utils.py

```diff
@@ -463,3 +465,166 @@ def _validate_expected_dep_file_contents(dep_contents_file: Path, output_dir: Pa
+    """Check if any actual code was affected by any of the changes.
+
+    Does this by checking if a change is in watched subtree.
+
+    >>> is_testable_code(Path('cachi2/core/config.py'))
+    True
+    >>> is_testable_code(Path('tests/integration/test_gomod.py'))
+    True
+    >>> is_testable_code(Path('tests/integration/utils.py'))
+    True
+    >>> is_testable_code(Path('tests/integration/conftest.py'))
+    True
+    >>> is_testable_code(Path('README.md'))
+    False
+    >>> is_testable_code(Path('reqruiements.txt'))
+    False
+    """
+    return any(c.is_relative_to(p) for p in paths_to_code)
+
+
+def select_testable_changes(changes: tuple[Path, ...]) -> tuple[Path, ...]:
+    """Weed out changes that cannot be tested.
+
+    If a change is outside of paths to testable code it is dropped as if
+    it never happened. Any file or module outside of watched directories
+    will be rejected.
+    """
+    return tuple(c for c in changes if is_testable_code(c))
+
+
+def just_some_package_managers_were_affected_by(changes: tuple[Path, ...]) -> bool:
+    """Check that just package managers were affected.
+
+    If any code outside of package managers subtree was affected or if a module
+    shared by package managers was affected will return False.
+
+    >>> just_some_package_managers_were_affected_by((Path('tests/integration/test_pip.py'),))
+    True
+    >>> c = Path('cachi2/core/package_managers/pip.py'), Path('tests/integration/test_pip.py')
+    >>> just_some_package_managers_were_affected_by(c)
+    True
+    >>> c = (Path('cachi2/core/package_managers/rpm/main.py'),)
+    >>> just_some_package_managers_were_affected_by(c)
+    True
+    >>> c = Path('cachi2/core/package_managers/gomod.py'), Path('tests/integration/test_pip.py')
+    >>> just_some_package_managers_were_affected_by(c)
+    True
+    >>> c = Path('cachi2/core/package_managers/general.py'), Path('tests/integration/test_pip.py')
+    >>> just_some_package_managers_were_affected_by(c)
+    False
+    >>> c = (Path('cachi2/core/package_managers/general.py'),
+    ...      Path('cachi2/core/package_managers/pip.py'))
+    >>> just_some_package_managers_were_affected_by(c)
+    False
+    >>> c = Path('tests/integration/utils.py'), Path('tests/integration/test_pip.py')
+    >>> just_some_package_managers_were_affected_by(c)
+    False
+    >>> c = Path('cachi2/core/package_managers/pip.py'), Path('tests/integration/utils.py')
+    >>> just_some_package_managers_were_affected_by(c)
+    False
+    >>> just_some_package_managers_were_affected_by((Path('cachi2/core/utils.py'),))
+    False
+    >>> c = (Path('cachi2/core/package_managers/general.py'),)
+    >>> just_some_package_managers_were_affected_by(c)
+    False
+    """
+    return all(affects_pm(c) for c in changes)
+
+
+def must_test_all() -> bool:
+    return os.getenv("CACHI2_RUN_ALL_INTEGRATION_TESTS", "false").lower() == "true"
+
+
+def determine_integration_tests_to_run_for(ref: str = "HEAD") -> Any:
```

Comment:
This is an incredibly confusing name, because the function does the exact opposite - it doesn't determine which tests to run, it returns the ones that need to be skipped.

Additionally, this should be typed as returning a `Set`, not `Any`.